### PR TITLE
Media 2020 queries.

### PR DESCRIPTION
**Pull request by smatei**
More media 2020 queries.

Progress on #900

### CODE REVIEW

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  var images = JSON.parse($._Images) || [];
+  return images.map(({url, naturalHeight, naturalWidth, width, height}) => ({url, naturalHeight: Number.parseInt(naturalHeight) || 0, naturalWidth: Number.parseInt(naturalWidth) || 0, width: Number.parseInt(width) || 0, height: Number.parseInt(height) || 0}));
+} catch (e) {}
+return null;
+''';
+
+SELECT
+  a.client,
+  imageType,
+  count(0) as count,
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:16]:
```suggestion
  count(0) AS count,
```

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  SELECT
+    _TABLE_SUFFIX AS client,
+    p.url AS page,
+    image.url AS url,
+    image.width width,
+    image.height height,
+    image.naturalWidth naturalWidth,
+    image.naturalHeight naturalHeight,
+    ifnull(image.width, 0) * ifnull(image.height, 0) pixels,
+    ifnull(image.naturalWidth, 0) * ifnull(image.naturalHeight, 0) naturalPixels
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:43]:
```suggestion
    image.width AS width,
    image.height AS height,
    image.naturalWidth AS naturalWidth,
    image.naturalHeight AS naturalHeight,
    IFNULL(image.width, 0) * IFNULL(image.height, 0) AS pixels,
    IFNULL(image.naturalWidth, 0) * IFNULL(image.naturalHeight, 0) AS naturalPixels
```

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  FROM
+    `httparchive.pages.2020_08_01_*` p
+    CROSS JOIN UNNEST(getImages(payload)) AS image
+  WHERE
+    image.naturalHeight > 0
+    AND image.naturalWidth > 0
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:49]:
```suggestion
    image.naturalHeight > 0 AND
    image.naturalWidth > 0
```

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+--  LIMIT 1000
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:50]: Do we need this? If not can we remove?

```suggestion
```

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+) a
+LEFT JOIN
+(
+    SELECT
+      client,
+      page,
+      url,
+      nullif(if(regexp_contains(mimetype, r'(?i)^application|^applicaton|^binary|^image$|^multipart|^media|^$|^text/html|^text/plain|\d|array|unknown|undefined|\*|string|^img|^images|^text|\%2f|\(|ipg$|jpe$|jfif'), format, lower(regexp_replace(regexp_replace(mimetype, r'(?is).*image[/\\](?:x-)?|[\."]|[ +,;]+.*$', ''), r'(?i)pjpeg|jpeg', 'jpg'))), '') as imageType,
+      respSize as bytes
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:59]:
```suggestion
      NULLIF(if(REGEXP_CONTAINS(mimetype, r'(?i)^application|^applicaton|^binary|^image$|^multipart|^media|^$|^text/html|^text/plain|\d|array|unknown|undefined|\*|string|^img|^images|^text|\%2f|\(|ipg$|jpe$|jfif'), format, LOWER(regexp_replace(regexp_replace(mimetype, r'(?is).*image[/\\](?:x-)?|[\."]|[ +,;]+.*$', ''), r'(?i)pjpeg|jpeg', 'jpg'))), '') AS imageType,
      respSize AS bytes
```

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      and (format <> '' OR mimetype like 'image%')
+
+      # many image/gifs are really beacons with 1x1 pixel, but svgs can get caught in the mix
+      and (respSize > 1500 OR regexp_contains(mimetype, r'svg'))
+
+      # strip favicon requests
+      and format <> 'ico'
+
+      # strip video mimetypes and other favicons
+      and not regexp_contains(mimetype, r'video|ico')
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:77]:
```suggestion
      # many 404s and redirects show up as image/gif
      status = 200 AND

      # we are trying to catch images. WPO populates the format for media but it uses a file extension guess.
      #So we exclude mimetypes that aren't image or where the format couldn't be guessed by WPO
     (format <> '' OR mimetype LIKE 'image%') AND

      # many image/gifs are really beacons with 1x1 pixel, but svgs can get caught in the mix
     (respSize > 1500 OR REGEXP_CONTAINS(mimetype, r'svg')) AND

      # strip favicon requests
     format <> 'ico' AND

      # strip video mimetypes and other favicons
      NOT REGEXP_CONTAINS(mimetype, r'video|ico')
```

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+-- limit 1000
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:78]: Delete this?

```suggestion
```

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+) b
+ON (b.client = a.client AND a.page = b.page AND a.url = b.url)
+
+WHERE
+  naturalPixels > 0
+  AND bytes > 0
+  AND imageType in ('jpg', 'png', 'webp', 'gif', 'svg')
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:85]:
```suggestion
  naturalPixels > 0 AND
  bytes > 0 AND
  imageType in ('jpg', 'png', 'webp', 'gif', 'svg')
```

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+GROUP BY
+  client,
+  imageType
+ORDER BY
+  client desc;
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:90]:
```suggestion
  client DESC
  imageType DESC;
```

File: sql/2020/05_Media/img_covid_in_path.sql [hunk 10]
```diff
@@ -0,0 +1,23 @@
+#standardSQL
+# images with covid in path - average size
+SELECT
+  client,
+  LOWER(ext) AS ext,
+  COUNT(0) AS ext_count,
+  COUNTIF(REGEXP_CONTAINS(LOWER(url), r'[^/]*?[:]//[^/]*?/.*?covid')) as ext_count_covid,
+  SAFE_DIVIDE(sum(respSize), COUNT(0)) as avg_size,
```
Comment (bazzadp) [inline sql/2020/05_Media/img_covid_in_path.sql:8]:
```suggestion
  COUNTIF(REGEXP_CONTAINS(LOWER(url), r'[^/]*?[:]//[^/]*?/.*?covid')) AS ext_count_covid,
  SAFE_DIVIDE(SUM(respSize), COUNT(0)) AS avg_size,
```

File: sql/2020/05_Media/img_covid_in_path.sql [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  SAFE_DIVIDE(sum(if(REGEXP_CONTAINS(LOWER(url), r'[^/]*?[:]//[^/]*?/.*?covid'), respSize, 0)), COUNTIF(REGEXP_CONTAINS(LOWER(url), r'[^/]*?[:]//[^/]*?/.*?covid'))) AS avg_size_covid
```
Comment (bazzadp) [inline sql/2020/05_Media/img_covid_in_path.sql:9]:
```suggestion
  SAFE_DIVIDE(SUM(if(REGEXP_CONTAINS(LOWER(url), r'[^/]*?[:]//[^/]*?/.*?covid'), respSize, 0)), COUNTIF(REGEXP_CONTAINS(LOWER(url), r'[^/]*?[:]//[^/]*?/.*?covid'))) AS avg_size_covid
```

File: sql/2020/05_Media/img_covid_in_path.sql [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+FROM
+  `httparchive.almanac.requests`
+WHERE
+  date = '2020-08-01' and
```
Comment (bazzadp) [inline sql/2020/05_Media/img_covid_in_path.sql:13]:
```suggestion
  date = '2020-08-01' AND
```

File: sql/2020/05_Media/img_covid_in_path.sql [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+  type = 'image'
+GROUP BY
+  client,
+  ext
+HAVING
+  ext_count_covid > 100
+ORDER BY
+  client,
+  ext_count_covid DESC;
+
```
Comment (bazzadp) [inline sql/2020/05_Media/img_covid_in_path.sql:23]:
```suggestion
```

File: sql/2020/05_Media/pixel_volume.sql [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+return null;
+''';
+
+SELECT
+  client,
+  count(0) as count,
+  any_value(viewportHeight) viewportHeight,
+  any_value(viewportWidth) viewportWidth,
+  any_value(dpr) dpr,
+  any_value(viewportHeight) * any_value(viewportWidth) displaypx,
```
Comment (bazzadp) [inline sql/2020/05_Media/pixel_volume.sql:29]:
```suggestion
  count(0) AS count,
  any_value(viewportHeight) AS viewportHeight,
  any_value(viewportWidth) AS viewportWidth,
  any_value(dpr) AS dpr,
  any_value(viewportHeight) * any_value(viewportWidth) AS displaypx,
```

File: sql/2020/05_Media/pixel_volume.sql [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    _TABLE_SUFFIX AS client,
+    url as page,
+    getCssPixels(json_extract_scalar(payload, '$._Images')) as cssPixels,
+    getNaturalPixels(json_extract_scalar(payload, '$._Images')) as naturalPixels,
+    CAST(json_extract_scalar(payload, '$._smallImageCount') AS Int64) as smallImageCount,
+    CAST(json_extract_scalar(payload, '$._bigImageCount') AS Int64) as bigImageCount,
+    CAST(json_extract_scalar(payload, '$._image_total') AS Int64) as imageBytes,
+    CAST(json_extract_scalar(json_extract_scalar(payload, '$._Dpi'), '$.dppx') AS Float64) as dpr,
+    CAST(json_extract_scalar(json_extract_scalar(payload, '$._Resolution'), '$.absolute.height') AS Float64) as viewportHeight,
+    CAST(json_extract_scalar(json_extract_scalar(payload, '$._Resolution'), '$.absolute.width') AS Float64) as viewportWidth
```
Comment (bazzadp) [inline sql/2020/05_Media/pixel_volume.sql:57]:
```suggestion
    url AS page,
    getCssPixels(JSON_EXTRACT_SCALAR(payload, '$._Images')) AS cssPixels,
    getNaturalPixels(JSON_EXTRACT_SCALAR(payload, '$._Images')) AS naturalPixels,
    CAST(JSON_EXTRACT_SCALAR(payload, '$._smallImageCount') AS Int64) AS smallImageCount,
    CAST(JSON_EXTRACT_SCALAR(payload, '$._bigImageCount') AS Int64) AS bigImageCount,
    CAST(JSON_EXTRACT_SCALAR(payload, '$._image_total') AS Int64) AS imageBytes,
    CAST(JSON_EXTRACT_SCALAR(json_extract_scalar(payload, '$._Dpi'), '$.dppx') AS Float64) AS dpr,
    CAST(JSON_EXTRACT_SCALAR(json_extract_scalar(payload, '$._Resolution'), '$.absolute.height') AS Float64) AS viewportHeight,
    CAST(JSON_EXTRACT_SCALAR(json_extract_scalar(payload, '$._Resolution'), '$.absolute.width') AS Float64) AS viewportWidth
```

File: sql/2020/05_Media/pixel_volume.sql [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+  FROM
+    `httparchive.pages.2020_08_01_*`
+--  LIMIT 1000
```
Comment (bazzadp) [inline sql/2020/05_Media/pixel_volume.sql:60]:
```suggestion
```

File: sql/2020/05_Media/pixel_volume.sql [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+)
+WHERE
+  # it appears the _Images array is populated only from <img> tag requests and not CSS or favicon
+  # likewise the bigImageCount and smallImageCount only track images > 100,000 and < 10,000 respectively.
+  # Meaning images between 10KB and 100KB won't show up in the count
+  # https://github.com/WPO-Foundation/webpagetest/blob/master/www/breakdown.inc#L95
+  cssPixels > 0 AND naturalPixels > 0
+  AND (smallImageCount > 0 OR bigImageCount > 0)
```
Comment (bazzadp) [inline sql/2020/05_Media/pixel_volume.sql:68]:
```suggestion
  cssPixels > 0 AND
  naturalPixels > 0 AND
  (smallImageCount > 0 OR bigImageCount > 0)
```

**Review by bazzadp — CHANGES_REQUESTED** [review]
Not looked too closely at the queries but lots of formatting nits

BTW use the "Add suggestions to batch" in Files tab to allow you to commit in one go which is a lot quicker.

File: sql/2020/05_Media/bytes_per_pixel.sql (hunk 4 above)
Comment (smatei) [reply sql/2020/05_Media/bytes_per_pixel.sql:50]: @barrypollard I copy pasted the sql from 2019, along with the formatting issues, I will fix it. As for the limit, I guess we can remove this.

File: sql/2020/05_Media/bytes_per_pixel.sql [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+      # strip video mimetypes and other favicons
+      NOT REGEXP_CONTAINS(mimetype, r'video|ico')
+) b
+ON (b.client = a.client AND a.page = b.page AND a.url = b.url)
+
+WHERE
+  naturalPixels > 0 AND
+  bytes > 0 AND
+  imageType in ('jpg', 'png', 'webp', 'gif', 'svg')
```
Comment (bazzadp) [inline sql/2020/05_Media/bytes_per_pixel.sql:83]:
```suggestion
  imageType IN ('jpg', 'png', 'webp', 'gif', 'svg')
```

File: sql/2020/05_Media/pixel_volume.sql [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+  # likewise the bigImageCount and smallImageCount only track images > 100,000 and < 10,000 respectively.
+  # Meaning images between 10KB and 100KB won't show up in the count
+  # https://github.com/WPO-Foundation/webpagetest/blob/master/www/breakdown.inc#L95
+  cssPixels > 0 AND
+  naturalPixels > 0 AND
+  (smallImageCount > 0 OR bigImageCount > 0)
+GROUP BY
+  client
+ORDER BY
+  client desc
```
Comment (bazzadp) [inline sql/2020/05_Media/pixel_volume.sql:72]:
```suggestion
  client DESC
```

**Review by bazzadp — APPROVED** [review]

**Comment by bazzadp** [comment]
> I copy pasted the sql from 2019, along with the formatting issues, I will fix it. As for the limit, I guess we can remove this.

Yeah we've tried to be a bit more disciplined this year! 😁 Thanks for the fixes. Spotted another couple that I fixed.

This good to merge? I'm happy if you are but still in Draft so didn't want to merge.

**Comment by smatei** [comment]
> 
> 
> > I copy pasted the sql from 2019, along with the formatting issues, I will fix it. As for the limit, I guess we can remove this.
> 
> Yeah we've tried to be a bit more disciplined this year! 😁 Thanks for the fixes. Spotted another couple that I fixed.
> 
> This good to merge? I'm happy if you are but still in Draft so didn't want to merge.

The data is in the spreadsheet, the authors said "thank you". I guess we can merge. Thanks.

**Pull request merged**